api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -181,6 +181,49 @@
         return getOptionalValue(propertyName, arrayType).map(Arrays::asList);
     }
 
+    /**
+     * Return the resolved configuration properties type with the specified prefix. 
+     * The type declaration can be annotated with 
+     * {@linkplain org.eclipse.microprofile.config.inject.ConfigProperties @ConfigProperties}. 
+     * If the type is annotated with {@linkplain org.eclipse.microprofile.config.inject.ConfigProperties @ConfigProperties}, 
+     * the prefix supplied in this method overrides the prefix associated with the annotation 
+     * {@linkplain org.eclipse.microprofile.config.inject.ConfigProperties @ConfigProperties}.
+     * 
+     * @param <T> 
+     *              The Class Type
+     * @param configProperties
+     *              The class that contains a number of fields that maps to corresponding configuration properties. 
+     *              The type declaration can be annotated with 
```

Comment:
Can be, or must be?

Reply:
It `can`. It's tricky, because you do need the annotation to be able to CDI scan it, but if you are using the Programmatic API, you don't need it at all, unless you want to set a prefix.

But I do see it may cause some confusion and cause inconsistencies, so maybe we should force to be always annotated?

Reply:
I have been there. At first, I tried to push the pattern of making the annotation mandatory. After talking with @radcortez @jbee, for programmtic lookup, it is more boilerplater if no prefix is to be supplied, suggested by @radcortez . Technicially speaking, the annotation does not need to be there. Thinking it further. I think for code reability, I think it is better to force this annotation to be mandatory.

Reply:
@jbee @radcortez @ljnelson please give you vote here if you want to make `@ConfigProperties` mandatory on the classes for injection and programmatically lookup. Personally, I think this will make code more readable and less magic.

Reply:
(I think what is bothering me about this entire feature is that it is reinventing CDI producers, sort of, but with slightly less rigor, and seemingly requiring reflection, though I haven't fully thought that part out.)  I haven't looked, but I hope `@ConfigProperties` is at least a `@Qualifier` and, if so, yes, I would expect it to be mandatory.

Reply:
Making it mandatory for CDI reasons seems backwards to me. CDI support is a feature, not the feature. For programmatic lookup no annotation is required functionally so I don't see why we would make it mandatory. This would just open up for things not working at runtime because of a missing annotation. I personally rather have defined working behaviour that might be useful to someone than another case that throws an exception.

Reply:
@ljnelson we do have a programmatic API, so we need to support it somehow to provide a similar experience between the CDI API and the programmatic API. Not everyone uses CDI.

Do you have any ideia / suggestion on how to support this in the programmatic API?

Reply:
Just use `Converter`s (the other pseudo-producer mechanism in this specification) which has the added feature that you don't get property drift.

Consider a `host`/`port` pair of configuration properties from a `ConfigSource` backed by a database.  Let's say you use this `@ConfigProperties` feature to create something with a `getHost()` and a `getPort()` method.  As you're assembling this POJO, the port changes, and the combination no longer makes sense.  Now you produce a POJO with inconsistent information ("property drift").  If, instead, you make the host/port pair atomic by encoding them together in a `String` and then using a `Converter` to produce the kind of POJO you want, you (a) achieve everything this feature is proposing while (b) preventing property drift and (c) keeping your scope narrow.

Another alternative is allowing for a special "root" property name that, when asked for, retrieves the "whole thing", and then allowing `Converter`s to work on that "whole thing" in some fashion.  This is sort of like Dropwizard configuration: if you slurp the "whole thing" in and then use Jackson or something similar, you get a POJO for the whole configuration.  Works great; dead simple; no fancy business needed; keeps you from having to invent yet another `String`s-to-`Object` facility.  (Suffers from property drift.)

At one point, this specification deliberately said in the interest of simplicity: we're going to treat configuration atoms as strings, full stop (hence `Properties`-like syntax).  You can convert them into small typed objects for mostly convenience, full stop.  If you want something bigger, assemble it yourself.  I feel like this spirit (which was one of the good things about the early specification) is getting lost.

(@jbee Given MicroProfile's stance towards CDI ("CDI first"), I assume CDI is more than just "a feature".  I do understand you need to support a programmatic API because CDI doesn't let you use CDI features to configure CDI as CDI is coming up.)

Reply:
@ljnelson you completely lost me here. I don't really understand what you are trying to convene. You talked a lot about dynamic changes and mismatch. This feature is not targetting for dynamic changes. If you worry about the mismatching, you either cache them in the implementation. 

@radcortez suggests making `ConfigProperties` annotation mandatory while @jbee wants the current behaviour which means in programmatic lookup for getConfigProperty() the type does not need to have the annotation present. What is your opinion?

Reply:
I'm not just completely sure how a `Converter` can help here. The `Converter` has a well defined API. It takes a single value and converts it to the target type. In here we are talking about grouping and retrieving multiple values. `Converter` will still be used to convert defined single values in the `ConfigProperties`. @ljnelson can you maybe provide some code samples (or pseudocode) on how the Converter API will be used in this case?

@Emily-Jiang I do prefer that is not mandatory yes, but I also see consistency issues. Let's just say that the annotation is optional in the programmatic API.

Reply:
@Emily-Jiang Sorry about that; I must have been unclear.  Let's try again.

Some `ConfigSource`s are provided by the user. Some are provided by the system. [Some cache; some do not](https://github.com/eclipse/microprofile-config/issues/370). This `@ConfigProperties` feature will be provided by the MicroProfile implementation, not the user. This feature can get values from multiple `ConfigSource`s, since it retrieves multiple properties. Therefore there is no guarantee of "transactionality": no guarantee that all the multiple properties gathered together by this feature are consistent, as each `ConfigSource` can change over time while the gathering is in progress.  Either we say transactionality is completely irrelevant to configuration, or it is not.  If it is not, then we need some way to ensure at least a pseudo-atomic transaction.  This whole feature seems like it's at the wrong level; I'd expect it to be in the application.  I hope this is clearer.

Reply:
@radcortez Pretend for a moment there is a "special" reserved property name.  Let's call it "`.`" just so I can talk about it.  Pretend further that when someone asks for "`.`" via `Config.getValue()`, she gets the entire configuration in Java Properties-like format, but as a single `String`.  Now envision a `Converter` that accepts such a `String` and converts it to a POJO.

This is roughly similar to how DropWizard configuration works.

Reply:
> @radcortez Pretend for a moment there is a "special" reserved property name. Let's call it "`.`" just so I can talk about it. Pretend further that when someone asks for "`.`" via `Config.getValue()`, she gets the entire configuration in Java Properties-like format, but as a single `String`. Now envision a `Converter` that accepts such a `String` and converts it to a POJO.
> 
> This is roughly similar to how DropWizard configuration works.

I get the idea, but what about performance wise? Potentially, you will need to select every config value to build such a String, since you don't know what is required. You could introspect the target object and provide some filtering, but then why would you need the Converter?

Reply:
Thank you @ljnelson for your explaination! Now I undestand what you saying. ConfigProperties is a way to gather properties. At the moment, what you described could exist if you are trying to build url from host and port and they are dynamic. I don't see why this introduce this problem. On the contrary, I think it helps the problem as host and port will be looked up pretty at the same time instead of scatter around and they were looked up at a completely different time.

Reply:
I guess then I've provided as much feedback as I can.  I do urge you to look at DropWizard configuration for further inspiration.

Reply:
I'll have a look.

Reply:
Hi @ljnelson I had a look, and followed this page: https://www.dropwizard.io/en/latest/manual/core.html#configuration

Please correct me if I'm wrong. Dropwizard provides you a YAML representation of the configuration as is, meaning overrides values already applied, which in fact translate to a truly immutable configuration representation. Then Jackson acts a converter mechanism to convert the YAML (or sections) in the objects that you expect.

If my assessment is correct, on our side, providing a full configuration representation may be costly, since this involves querying all the registered sources (which we don't know how many are involved). Dropwizard seem to limit to the file, system properties and env - I couldn't find a way to register additional sources. We could limit this by analyzing the target object and retrieve only what is required to build it.

Passing this data structure to the `Converter`, will certainly allow us to be able to convert it to whatever POJO the Converter is handling, but this does require the `Converter` to support nested conversions, by calling other converters etc. We now enter into the discussion about having a separate Converter spec, since our conversion API is limited to the use cases defined by the specification. I've done some work in the area and I was trying to come up with a way to separate Conversion from Config. Here are some of the efforts in that area:

https://github.com/smallrye/smallrye-converters/
https://github.com/smallrye/smallrye-config/pull/265

Anyway, looking into the current state of the proposal, nothing says that it cannot be implemented with a `Converter`. The new methods and annotation are mainly there for the `prefix`, (or we can think about this in a way to filter the configuration properties).  

@struberg did suggest to use the existent `getValue(String, Class)` which effectively suggests that a `Converter` would be used. I did push to have it as a separate API, mainly because the POJO may contain all the lookup information and it would look weird to me to call `getValue("", Pojo.class)` or `getValue(".", Pojo.class)` or even `getValue(null, Pojo.class`). I guess we could all a `getValue(Class)` method for that case, but personally it would look weird to me, because you are retrieving a set of values and not just a single value. Then in the `getValue(String, Class) method, the `String` parameter would have different semantics depending on how the method is called.

Sorry, this was a long post, but in conclusion I believe that nothing stops the implementation to use Converters, and we can certainly review / change the API to accommodate it. Please, let me know your thoughts.

Reply:
(I believe DropWizard also applies overrides from System properties somehow but I'm not sure off the top of my head, so the configuration is not necessarily completely immutable.  The main point is: they start from POJOs.  We seem confused: are we primarily a system that makes an atomic request for a single scalar ("please atomically get the value for `frobnicator.host`"), or are we primarily a system that makes a pseudo-atomic request for groups of scalars ("please as atomically as possible get the values for `frobnicator.host` and `frobnicator.port`")?  DropWizard says: "please get everything and turn it into a configuration object all at once").  We of course already have the means of iterating over all `ConfigSource`s: that's what `Config` is for.  All we lack is a means of specifying "everything".)

Another thing worth bearing in mind is that MicroProfile Config is slowly and erratically rehashing JNDI.  One of the first things that JNDI had to figure out was: are nodes simple nodules with children, or are they nodules with children and values?  MicroProfile Config to its credit in the early days took a very opinionated stand and said: everything is a scalar.  Now here we are reinventing `ObjectFactory` concepts and we're up against that decision again.

Coming back to the issue: it's not that I think `Converter`s are a great API or even necessarily the right tool for this job.  It's that I feel that this feature doesn't "fit" with the original spirit of the specification, which was: we will give you a way to get `String` scalars out of a composable configuration system, together with some simple `String`-to-number conversions etc. and…that's it.  We're done.  CDI is much better at producing objects than this specification will ever be.  JavaBeans is better at it.  Some weird hybrid of the two seems like a very, very strange thing indeed to put into an implementation, let alone a specification.  Is this feature necessary?  Can't we just let grouping of scalars (since they're going to be non-transactional anyway) happen in the application?  Why must it be specified?

Reply:
@ljnelson this feature is nothing more than gathering the config properties into one pojo and nothing more. In you sample of frobnicator.host, at the moment you use `@Inject @ConfigProperty(name="frobinator.host")` and then `@Inject @ConfigProperty(name="frobinator.port")`. With this proposal, you just need to specify one ConfigProperties. It is much simplier for end users to use. As for implementation, you could provide a converter for the ConfigProperties class. In the converter itself, it will call into other converters to resolve individual fields. By the way, [DeltaSpike Config](https://deltaspike.apache.org/documentation/configuration.html) and Spring config (https://www.baeldung.com/configuration-properties-in-spring-boot) both support this feature. I am not sure why you think it should not be done.

Reply:
Then I'll get out of the way.